miner/worker_builder.go

```diff
@@ -0,0 +1,406 @@
+)
+
+// fillTransactions retrieves the pending bundles and transactions from the txpool and fills them
+// into the given sealing block. The selection and ordering strategy can be extended in the future.
+func (w *worker) fillTransactionsAndBundles(interrupt *atomic.Int32, env *environment) error {
+	// TODO will remove after fix txpool perf issue
+	if interrupt != nil {
+		if signal := interrupt.Load(); signal != commitInterruptNone {
+			return signalToErr(signal)
+		}
+	}
+
+	bundles := w.eth.TxPool().PendingBundles(env.header.Number.Uint64(), env.header.Time)
+
+	// if no bundles, not necessary to fill transactions
+	if len(bundles) == 0 {
+		return errors.New("no bundles in bundle pool")
+	}
+
+	txs, _, err := w.generateOrderedBundles(env, bundles)
+	if err != nil {
+		log.Error("fail to generate ordered bundles", "err", err)
+		return err
+	}
+
+	if err = w.commitBundles(env, txs, interrupt); err != nil {
+		log.Error("fail to commit bundles", "err", err)
+		// if commit bundles failed, set profit is 0, discard all fill bundles
+		env.profit.Set(big.NewInt(0))
+		return err
+	}
+	log.Info("fill bundles", "bundles_count", len(bundles))
+
+	start := time.Now()
+	pending := w.eth.TxPool().Pending(true)
+	packFromTxpoolTimer.UpdateSince(start)
+	log.Debug("packFromTxpoolTimer", "duration", common.PrettyDuration(time.Since(start)), "hash", env.header.Hash())
+
+	// Split the pending transactions into locals and remotes.
+	localTxs, remoteTxs := make(map[common.Address][]*txpool.LazyTransaction), pending
+
+	// Fill the block with all available pending transactions.
+	start = time.Now()
+	if len(localTxs) > 0 {
+		txs := newTransactionsByPriceAndNonce(env.signer, localTxs, env.header.BaseFee)
+		if err := w.commitTransactions(env, txs, interrupt); err != nil {
+			return err
+		}
+	}
+	if len(remoteTxs) > 0 {
+		txs := newTransactionsByPriceAndNonce(env.signer, remoteTxs, env.header.BaseFee)
+		if err := w.commitTransactions(env, txs, interrupt); err != nil {
+			return err
+		}
+	}
+	commitTxpoolTxsTimer.UpdateSince(start)
+	log.Debug("commitTxpoolTxsTimer", "duration", common.PrettyDuration(time.Since(start)), "hash", env.header.Hash())
+	log.Info("fill bundles and transactions done", "total_txs_count", len(env.txs))
+	return nil
+}
+
+func (w *worker) commitBundles(
+	env *environment,
+	txs types.Transactions,
+	interrupt *atomic.Int32,
+) error {
+	gasLimit := env.header.GasLimit
+	if env.gasPool == nil {
+		env.gasPool = new(core.GasPool).AddGas(gasLimit)
+	}
+	var coalescedLogs []*types.Log
+
+	for _, tx := range txs {
+		if interrupt != nil {
+			if signal := interrupt.Load(); signal != commitInterruptNone {
+				return signalToErr(signal)
+			}
+		}
+		// If we don't have enough gas for any further transactions then we're done.
+		if env.gasPool.Gas() < params.TxGas {
+			log.Trace("Not enough gas for further transactions", "have", env.gasPool, "want", params.TxGas)
+			break
```

Comment:
Will this break happen in the middle of a bundle, which results in partial txns of a bundle included in block?

Reply:
Yes, it is incorrect. Fixed, please review again.